packages/hash/backend-utils/src/search/opensearch.ts

```diff
@@ -162,4 +222,188 @@ export class OpenSearch extends DataSource implements SearchAdapter {
 
     return { hits };
   }
+
+  private async clearScrollAndReturnUndefined(cursor: string) {
+    await this.client.clear_scroll({ scroll_id: cursor });
+    return undefined;
+  }
+
+  /**
+   * Decide whether to increment the search cursor or to clear it.
+   * If there are no more pages left (we've seen all `total` hits) then clear
+   *
+   * @param hitCount number of hits in the result
+   * @param seenCount number of hits seen so far (not including the newly seen)
+   * @param total  number of hits for the whole search
+   * @param openSearchCursor cursor used to receive current result
+   * @param nextOpenSearchCursor cursor to use after the current result
+   * @returns a search cursor that contain more data or `undefined`
+   */
+  private async incrementOrClearOpenSearchCursor(params: {
+    pageSize: number;
+    hitCount: number;
+    seenCount: number;
+    total: number;
+    openSearchCursor: string;
+    nextOpenSearchCursor: string | undefined;
+  }) {
+    const {
+      hitCount,
+      nextOpenSearchCursor,
+      openSearchCursor,
+      seenCount,
+      total,
+      pageSize,
+    } = params;
+    const seenSoFar = seenCount + hitCount;
+
+    const endOfSearch =
+      hitCount === 0 ||
+      nextOpenSearchCursor === undefined ||
+      seenSoFar >= total;
+
+    if (endOfSearch) {
+      return await this.clearScrollAndReturnUndefined(openSearchCursor);
+    }
+
+    // A new cursor requires the page to have hits and a proper next cursor
+    const nextCursor =
+      nextOpenSearchCursor && seenSoFar < total
+        ? opaqueCursorToString({
+            seenCount: seenSoFar,
+            openSearchCursor: nextOpenSearchCursor,
+            pageSize,
+          })
+        : await this.clearScrollAndReturnUndefined(openSearchCursor);
+
+    // Only return a new search cursor if this is not the final page
+    // If there are not hits in the page, the newCursor will be undefiend as well.
+    return seenCount < total ? nextCursor : undefined;
+  }
+
+  /**
+   * Initiate paginated search with search parameters
+   * The current implementation makes use of OpenSearch Scroll,
+   * which stores search results in a context in the OpenSearch server.
+   *
+   * The cursor is kept alive for 10 minutes.
```

Comment:
If I've understood this correctly, this is a pretty serious limitation which we should discuss the consequences of and alternatives to - it'll make harder for API consumers to programmatically do work through pages of results, for example, if the work takes longer than 10 minutes. And prevent browser users from paging through results slowly (e.g. if they switch task for 10 minutes and then come back and hit the next page button).

I don't think we've spoken about `search_after` so it'd be good to understand the trade-offs involved.

Reply:
I believe `search_after` requires us to have sortable ids (I added a section to the cursor document on notion to address the potential before starting work on this).

It would require we do some of the paging logic ourselves, which shouldn't be too bad, and we wouldn't have to have a close-timer on pagination. It would definitely be beneficial to use `search_after` for our purposes instead of scroll, though.

That being said, we can increase this timeout if we want to. I am not sure what the limitations on AWS hosted OpenSearch is in regards to this.

Reply:
Cool, thank you. Presumably it would only require us to have sortable ids if we want to return a sort that is sorted by ids - which I don't think we ever want to do. We could use `createdAt` as a default sort instead, if the user doesn't specify any. 

Let's talk more in the next refinement session about this.

Reply:
By the way, using `search_after` would allow us to do pagination in both directions by changing sort ordering

Reply:
Even better!